### PR TITLE
Fix Issue 2393

### PR DESCRIPTION
**Pull request by openamiguel**
Fix for Issue #2393 (WrongOneof mishandles new style switch statement) based on PR #2371 and PR #2124.

### CODE REVIEW

**Comment by openamiguel** [comment]
@googlebot I signed it!

**Review by cushon — CHANGES_REQUESTED** [review]
Thanks for the fix! Can you also add a test?

**Comment by openamiguel** [comment]
Sure. There isn't any WrongOneofTest currently. Can I just focus on testing that this bug pattern works on new-style switch statements?

**Comment by cushon** [comment]
> There isn't any WrongOneofTest currently. Can I just focus on testing that this bug pattern works on new-style switch statements?

Sorry, we have an internal test for this, but it isn't available in open source to avoid the proto dependency in the maven build. Since there isn't already an existing test, this is fine.

**Review by cushon — APPROVED** [review]

**Comment by openamiguel** [comment]
@cushon thanks for the approval! Are we waiting for additional review?

**Comment by cushon** [comment]
> Are we waiting for additional review?

Yes, there's a separate internal review step before the PR is merged

**Pull request closed without merging**
